### PR TITLE
Upgrade/mern v2.4.0

### PR DESCRIPTION
**Pull request by mannyhenri**
See change.md for all changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,17 +13,28 @@ MERN is a scaffolding tool which makes it easy to build isomorphic apps using Mo
...
+```sh
   npm install -g mern-cli
   mern init your_new_app
   cd your_new_app
   npm install
   npm start
 ```
 
-**Note : Please make sure your MongoDB is running.** For MongoDB installation guide see [this](https://docs.mongodb.org/v3.0/installation/). Also `npm3` is required to install dependencies properly.
+**Note : Please make sure your MongoDB is running.** For MongoDB installation guide see [this](https://docs.mongodb.org/v3.0/installation/). Also `npm5` is required to install dependencies properly.
```
Comment (MrChristofferson) [inline README.md]: Are we encouraging the use of npm at 5 or 6? `changes.md` says 6 :)

File: .travis.yml [hunk 2]
```diff
@@ -1,11 +1,12 @@
 language: node_js
 node_js:
-  - '5'
-  - '4'
+  - '6'
+  - '8'
+  - '10'
 before_install:
   - 'npm install -g npm@latest'
```
Comment (MrChristofferson) [inline .travis.yml]: Should we specify a version here?

File: README.md (hunk 1 above)
Comment (mannyhenri) [reply README.md]: @mattlubner let me know how you'd prefer to proceed related to npm version, stick to 5 or use 6.

File: .travis.yml (hunk 2 above)
Comment (mannyhenri) [reply .travis.yml]: I'd suggest we specify 8 (LTS) but you guys let me know.

**Comment by mannyhenri** [comment]
Doesn't show here but all tests have passed on Travis: https://travis-ci.org/Hashnode/mern-starter

File: .travis.yml (hunk 2 above)
Comment (MrChristofferson) [reply .travis.yml]: I was actually referring to the `npm@latest` in this comment, so it goes right along with the one up there ^

Comment (MrChristofferson) [reply .travis.yml]: @mattlubner

Comment (mattlubner) [reply .travis.yml]: Hey @MrChristofferson @mannyhenri I chose those nodejs versions for testing in TravisCI as they're the ones actively being maintained (see: https://github.com/nodejs/Release#release-schedule). I was thinking we should test on whatever's actively being maintained, but what do you think? (note that I omitted v9 as it's not LTS and gets EOL'd next month)

As for the `npm` version, I totally agree, we should specify it explicitly here. Will do that shortly! 🙂

File: README.md (hunk 1 above)
Comment (mannyhenri) [reply README.md]: Just made the readme change.

File: Changes.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+- Fix test runner windows
+- Updated to webpack 3
+- Use .dockerignore to increase speed of building Docker image (no need to copy files to Docker build context).
+- Use npm@6 and add package-lock.json.
+- Update versions of nodejs used for testing in Travis (only LTS / actively maintained versions - no more node@4!!).
+- Use a single Dockerfile w/ build targets, to greatly speed up building of Docker image.
+- Swap default Docker environment from production to development, to improve DX.
+- Eliminate production-mode inclusion of devDependencies, to keep production Docker image as light as possible and lower production-mode memory usage.
+- Explicitly volume-map all files into Docker container when in dev-mode.
+- Restore mongodb port mapping, for test suite (temporary - this should be removed once the test suite properly uses an in-memory mock db).
```
Comment (mattlubner) [inline Changes.md]: Actually I think we can safely remove this now that the mockgoose setup has been fixed.

File: Dockerfile [hunk 4]
```diff
@@ -1,15 +1,27 @@
...
 
-COPY package.json /usr/src/app
+FROM base as development
+ENV NODE_ENV development
+COPY package.json package-lock.json ./
 RUN npm install
-COPY . /usr/src/app
+COPY .babelrc .eslintignore .eslintrc index.js nodemon.json webpack.config.babel.js webpack.config.dev.js webpack.config.prod.js webpack.config.server.js ./
+COPY client ./client
+COPY config ./config
```
Comment (mattlubner) [inline Dockerfile]: Haha so I'm totally nitpicking my own code here… we don't actually need `config/` in the Dockerfile, as it's meant for local machine dev'ing, right?

File: docker-compose-production.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      - db
+    ports:
+      - "${WEB_PORT:-80}:8000"
+    environment:
+      NODE_ENV: production
+      MONGO_URL: mongodb://db:27017/mern-starter
+  db:
+    image: mongo:latest
+    ports:
+      - "27017:27017"
```
Comment (mattlubner) [inline docker-compose-production.yml]: Should we map the mongodb port here? It's not necessary and could potentially conflict with users who have mongodb running locally.

(edit: I had thrown it in here when the test suite required a real mongodb instance to be running… it no longer does)

File: nodemon.json [hunk 6]
```diff
@@ -9,6 +9,7 @@
     "server",
     "Intl"
   ],
+  "legacyWatch": true,
```
Comment (mattlubner) [inline nodemon.json]: Using legacy watch is not ideal and we should avoid it, if possible.

**Comment by mattlubner** [comment]
Just removed my needs-approval. @mannyhenri would you like to review my changes?

**Pull request merged**